### PR TITLE
[JENKINS-60945] Improve inspection of commits

### PR DESCRIPTION
**Pull request by rathaya**
Implemented inspeciton of all repository commits.

https://issues.jenkins-ci.org/browse/JENKINS-60945?jql=project%20%3D%20JENKINS%20AND%20component%20%3D%20git-forensics-plugin

### CODE REVIEW

File: plugin/src/main/java/io/jenkins/plugins/forensics/git/miner/GitRepositoryMiner.java [hunk 1]
```diff
@@ -69,68 +78,100 @@ public RepositoryStatistics mine(final Collection<String> absoluteFileNames, fin
...
+                    files = getFilesFromCommit(repository, git, null, commits.get(i).getName());
+                }
+                else {
+                    if (i > 0) {
+                        files = getFilesFromCommit(repository, git, commits.get(i + 1).getName(),
+                                commits.get(i).getName());
+                    }
+                }
+                int finalI = i;
+                files.forEach(f -> fileStatistics.computeIfAbsent(f, builder::build)
```
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/forensics/git/miner/GitRepositoryMiner.java:125]: Wouldn't it be simpler to extract the files from the diff?

File: plugin/src/main/java/io/jenkins/plugins/forensics/git/miner/GitRepositoryMiner.java [hunk 2]
```diff
@@ -69,68 +78,100 @@ public RepositoryStatistics mine(final Collection<String> absoluteFileNames, fin
...
-                commits.forEach(c -> fileStatistics.inspectCommit(c.getCommitTime(), getAuthor(c)));
-                return fileStatistics;
+        private List<String> getFilesFromCommit(final Repository repository, final Git git, final String oldCommit,
+                final String newCommit) {
+            List<String> filePaths = new ArrayList<>();
+
+            try {
+                final List<DiffEntry> diffEntries = git.diff()
+                        .setOldTree(getTreeParser(repository, oldCommit))
+                        .setNewTree(getTreeParser(repository, newCommit))
```
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/forensics/git/miner/GitRepositoryMiner.java:140]: What happens if we do not set `oldTree` at all? Isn't the previous commit automatically used?

File: plugin/src/main/java/io/jenkins/plugins/forensics/git/miner/GitRepositoryMiner.java [hunk 3]
```diff
@@ -69,68 +78,100 @@ public RepositoryStatistics mine(final Collection<String> absoluteFileNames, fin
...
-                    .map(builder::build)
-                    .map(file -> analyzeHistory(repository, file))
-                    .collect(Collectors.toList());
-            statistics.addAll(fileStatistics);
-
-            logger.logInfo("-> created statistics for %d files", statistics.size());
-
+            List<String> files = new ArrayList<>();
+            Map<String, FileStatistics> fileStatistics = new HashMap<>();
+            Set<String> filesInHead = new FilesCollector(repository).findAllFor(repository.resolve(Constants.HEAD));
```
[the remark this answers is not in the document]
Comment (rathaya) [reply plugin/src/main/java/io/jenkins/plugins/forensics/git/miner/GitRepositoryMiner.java:113]: I use filesInHead to filter the deleted files.

I tried it first by comparing the oldPath and the newPath, but that takes more lines.

File: plugin/src/main/java/io/jenkins/plugins/forensics/git/miner/GitRepositoryMiner.java (hunk 2 above)
Comment (rathaya) [reply plugin/src/main/java/io/jenkins/plugins/forensics/git/miner/GitRepositoryMiner.java:140]: It does not seem to work without ``oldTree``, but I will take a closer look into this.

File: plugin/src/main/java/io/jenkins/plugins/forensics/git/miner/GitRepositoryMiner.java (hunk 1 above)
Comment (rathaya) [reply plugin/src/main/java/io/jenkins/plugins/forensics/git/miner/GitRepositoryMiner.java:125]: I'm not sure if I understand this point. I am extracting the file paths in ``getFilesFromCommit`` from the diff and create the corresponding FileStatistics here.

Comment (uhafner) [reply plugin/src/main/java/io/jenkins/plugins/forensics/git/miner/GitRepositoryMiner.java:127]: Sorry,  I thought you are iterating over `filesInHead`.

**Pull request merged**